staging/vhost-device-can/src/vhu_can.rs

```diff
@@ -0,0 +1,2181 @@
+                warn!(
+                    "%s(): Cut length from {} to 64\n",
+                    response.length.to_native()
+                );
+                can_rx.length = 64.into();
+            }
+        } else if response.length.to_native() > 8 {
+            warn!(
+                "%s(): Cut length from {} to 8\n",
+                response.length.to_native()
+            );
+            can_rx.length = 8.into();
+        }
+
+        can_rx.sdu.copy_from_slice(&response.sdu[0..64]);
+        CanController::print_can_frame(can_rx);
+
+        Ok(can_rx)
+    }
+
+    fn process_ctrl_requests(
+        &self,
+        requests: Vec<CanDescriptorChain>,
+        vring: &VringRwLock,
+    ) -> Result<bool> {
+        trace!("process_ctrl_requests");
+
+        if requests.is_empty() {
+            return Ok(true);
+        }
+
+        for desc_chain in requests {
+            let atomic_mem = self.mem.as_ref().unwrap().memory();
+
+            let mut reader = desc_chain
+                .clone()
+                .reader(&atomic_mem)
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            let mut writer = desc_chain
+                .clone()
+                .writer(&atomic_mem)
+                .map_err(|_| Error::DescriptorWriteFailed)?;
+
+            let request = reader
+                .read_obj::<VirtioCanCtrlRequest>()
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            // This implementation requires the CAN devices to be already in UP state
+            // before starting. This code does not trigger state changes [UP/DOWN] of
+            // the host CAN devices.
+            let response = match request.msg_type.into() {
+                VIRTIO_CAN_SET_CTRL_MODE_START => {
+                    trace!("VIRTIO_CAN_SET_CTRL_MODE_START");
+                    if self.controller.write().unwrap().ctrl_state == CAN_CS_STARTED {
+                        Ok(VIRTIO_CAN_RESULT_NOT_OK)
+                    } else {
+                        self.controller.write().unwrap().busoff = false;
+                        self.controller.write().unwrap().ctrl_state = CAN_CS_STARTED;
+                        Ok(VIRTIO_CAN_RESULT_OK)
+                    }
+                }
+                VIRTIO_CAN_SET_CTRL_MODE_STOP => {
+                    trace!("VIRTIO_CAN_SET_CTRL_MODE_STOP");
+                    if self.controller.write().unwrap().ctrl_state == CAN_CS_STOPPED {
+                        Ok(VIRTIO_CAN_RESULT_NOT_OK)
+                    } else {
+                        self.controller.write().unwrap().busoff = false;
+                        self.controller.write().unwrap().ctrl_state = CAN_CS_STOPPED;
+                        Ok(VIRTIO_CAN_RESULT_OK)
+                    }
+                }
+                _ => {
+                    trace!("Ctrl queue: msg type 0x{:?} unknown", request.msg_type);
+                    Err(Error::HandleCtrlUnknown)
+                }
+            }?;
```

Comment:
I wonder if, instead of returning with err, we should log the err, and then, write some response, e.g., VIRTIO_CAN_RESULT_NOT_OK. I do not know if it is correct to not return the descriptor chain that trigger the issue.